atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/expectations/KeyValueServiceDataTrackerTest.java

```diff
@@ -0,0 +1,278 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.impl.expectations;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.google.common.collect.ImmutableMap;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.transaction.api.expectations.ImmutableKvsCallReadInfo;
+import com.palantir.atlasdb.transaction.api.expectations.ImmutableTransactionReadInfo;
+import com.palantir.atlasdb.transaction.api.expectations.KvsCallReadInfo;
+import com.palantir.atlasdb.transaction.api.expectations.TransactionReadInfo;
+import org.junit.Test;
+
+public final class KeyValueServiceDataTrackerTest {
+    private static final TableReference TABLE_1 = TableReference.createWithEmptyNamespace("Table1");
+    private static final TableReference TABLE_2 = TableReference.createWithEmptyNamespace("Table2");
+    private static final TableReference TABLE_3 = TableReference.createWithEmptyNamespace("Table3");
+    private static final String KVS_METHOD_NAME_1 = "getRows";
+    private static final String KVS_METHOD_NAME_2 = "getAsync";
+    private static final String KVS_METHOD_NAME_3 = "get";
+    private static final String KVS_METHOD_NAME_4 = "getMetadataForTables";
+    private static final String KVS_METHOD_NAME_5 = "getMetadataForTable";
+    private static final long NO_BYTES_READ = 0L;
+    private static final long BYTES_READ_1 = 83L;
+    private static final long BYTES_READ_2 = 103L;
+    private static final long BYTES_READ_3 = 971L;
+
+    private final KeyValueServiceDataTracker tracker = new KeyValueServiceDataTracker();
+
+    @Test
+    public void noReadsTracksNothing() {
+        assertThat(tracker.getReadInfo()).isEqualTo(createTransactionReadInfo(0, 0));
+        assertThat(tracker.getReadInfoByTable()).isEmpty();
+    }
+
+    @Test
+    public void oneReadForTableIsTracked() {
+        tracker.readForTable(TABLE_1, KVS_METHOD_NAME_1, BYTES_READ_1);
+
+        TransactionReadInfo readInfo = createTransactionReadInfo(
+                BYTES_READ_1, 1, createMaximumBytesKvsCallInfo(BYTES_READ_1, KVS_METHOD_NAME_1));
+
+        assertThat(tracker.getReadInfo()).isEqualTo(readInfo);
+        assertThat(tracker.getReadInfoByTable()).containsExactlyEntriesOf(ImmutableMap.of(TABLE_1, readInfo));
+    }
+
+    @Test
+    public void multipleReadsForTableAreTracked() {
```

Comment:
Discussed offline: Partial goal of these tests is to test interleaving.
Consider: Have tests that order to method calls by say table name for readability. The oneRead is really subsumed by the multiple case, so you don't strictly need the oneRead cases (but I'm not opposed to having them around if you believe there is value).

Then, add a new test for that schedules a ton of concurrent tasks and assert that the value is expected. That way you're actually testing your concurrency control and interleaving.